pkg/defkinds/nodejs/builder.go

```diff
@@ -171,34 +170,65 @@ func getEnv(src llb.State, name string) string {
 
 func (h *NodeJSHandler) globalPackagesInstall(
 	state llb.State,
-	globalPackages map[string]string,
+	stageDef StageDefinition,
 	buildOpts builddef.BuildOpts,
 ) llb.State {
-	if len(globalPackages) == 0 {
+	if stageDef.GlobalPackages.Len() == 0 {
 		return state
 	}
 
-	pkgs := make([]string, 0, len(globalPackages))
-	for pkg, constraint := range globalPackages {
+	pkgs := make([]string, 0, stageDef.GlobalPackages.Len())
+	for pkg, constraint := range stageDef.GlobalPackages.Map() {
 		if constraint != "" && constraint != "*" {
 			pkg += "@" + constraint
 		}
 		pkgs = append(pkgs, pkg)
 	}
 
-	cmd := fmt.Sprintf("yarn global add %s", strings.Join(pkgs, " "))
-	runOpts := []llb.RunOption{
-		llbutils.Shell(cmd),
-		llb.User("1000"),
-		llb.WithCustomNamef("Run %s", cmd)}
+	runOpts := []llb.RunOption{llb.User("1000")}
+	if stageDef.PackageManager == pkgManagerYarn {
+		runOpts = append(runOpts,
+			llbutils.Shell("yarn global add "+strings.Join(pkgs, " ")),
+			llb.WithCustomName("Run yarn global add"))
+	} else {
+		runOpts = append(runOpts,
+			llbutils.Shell("npm install -g "+strings.Join(pkgs, " ")),
+			llb.WithCustomName("Run npm install"))
+	}
 
 	if buildOpts.IgnoreLayerCache {
 		runOpts = append(runOpts, llb.IgnoreCache)
 	}
 
+	runOpts, state = cacheMountOptForJSDeps(
+		runOpts, state, buildOpts, stageDef.PackageManager)
+
 	return state.Run(runOpts...).Root()
 }
 
+func cacheMountOptForJSDeps(
+	runOpts []llb.RunOption,
+	state llb.State,
+	buildOpts builddef.BuildOpts,
+	pkgMgr string,
+) ([]llb.RunOption, llb.State) {
+	if !buildOpts.WithCacheMounts {
```

Comment:
no need to dissociate the cache mount types ?

Reply:
Nope, the cache ID generated by `CacheMountOpt` should be enough :wink: